prov/gni/src/gnix_sep.c

```diff
@@ -0,0 +1,860 @@
+			if (ep->send_cq)
+				ep->tx_enabled = true;
+			if (ep->recv_cq)
+				ep->rx_enabled = true;
+		}
+
+		break;
+	case FI_GETFIDFLAG:
+	case FI_SETFIDFLAG:
+	case FI_ALIAS:
+	default:
+		return -FI_ENOSYS;
+	}
+err:
+	return ret;
+}
+
+static void __sep_destruct(void *obj)
+{
+	int i;
+	struct fid_domain *domain;
+	struct gnix_fid_ep *ep;
+	struct gnix_fid_domain *domain_priv;
+	struct gnix_fid_sep *sep = (struct gnix_fid_sep *) obj;
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	domain = sep->domain;
+	assert(domain != NULL);
+	domain_priv = container_of(domain, struct gnix_fid_domain, domain_fid);
+
+	_gnix_ref_put(domain_priv);
+
+	for (i = 0; i < sep->info->ep_attr->tx_ctx_cnt; i++) {
+		ep = container_of(sep->ep_table[i],
+				  struct gnix_fid_ep, ep_fid);
+		if (ep == NULL) {
+			continue;
+		}
+
+		if (ep->av) {
+			_gnix_ref_put(ep->av);
+		}
+	}
+
+	if (sep->ep_table)
+		free(sep->ep_table);
+	if (sep->tx_ep_table)
+		free(sep->tx_ep_table);
+	if (sep->rx_ep_table)
+		free(sep->rx_ep_table);
+	free(sep);
+}
+
+static int gnix_sep_close(fid_t fid)
+{
+	int ret = FI_SUCCESS;
+	int refs_held;
+	struct gnix_fid_sep *sep;
+	struct gnix_fid_trx *trx_priv;
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	switch (fid->fclass) {
+	case FI_CLASS_SEP:
+		sep = container_of(fid, struct gnix_fid_sep, ep_fid.fid);
+		refs_held = _gnix_ref_put(sep);
+		if (refs_held) {
+			GNIX_INFO(FI_LOG_CQ, "failed to fully close sep due to"
+				  " lingering references. refs=%i sep=%p\n",
+				  refs_held, sep);
+		}
+		break;
+	case FI_CLASS_TX_CTX:
+	case FI_CLASS_RX_CTX:
+		trx_priv = container_of(fid, struct gnix_fid_trx, ep_fid);
+		_gnix_ref_put(trx_priv);
+		break;
+	default:
+		return -FI_EINVAL;
+	}
+
+	return ret;
+}
+
+int gnix_sep_open(struct fid_domain *domain, struct fi_info *info,
+		 struct fid_ep **sep, void *context)
+{
+	struct gnix_fid_sep *sep_priv = NULL;
+	struct gnix_fid_domain *domain_priv = NULL;
+	int ret = FI_SUCCESS;
+	int n_ids = GNIX_SEP_MAX_CNT;
+	uint32_t cdm_id, cdm_id_base;
+	struct gnix_ep_name *name;
+	uint32_t name_type = GNIX_EPN_TYPE_UNBOUND;
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	if ((domain == NULL) || (info == NULL) || (sep == NULL) ||
+	    (info->ep_attr == NULL))
+		return -FI_EINVAL;
+
+	/*
+	 * TODO: use macro here
+	 */
+	if (!GNIX_EP_RDM_DGM(info->ep_attr->type))
+		return -FI_ENOSYS;
+
+	/*
+	 * check limits for rx and tx ctx's
+	 */
+
+	if ((info->ep_attr->tx_ctx_cnt > n_ids) ||
+	    (info->ep_attr->rx_ctx_cnt > n_ids))
+		return -FI_EINVAL;
+
+	domain_priv = container_of(domain, struct gnix_fid_domain, domain_fid);
+
+	sep_priv = calloc(1, sizeof(*sep_priv));
+	if (!sep_priv)
+		return -FI_ENOMEM;
+
+	sep_priv->type = info->ep_attr->type;
+	sep_priv->ep_fid.fid.fclass = FI_CLASS_SEP;
+	sep_priv->ep_fid.fid.context = context;
+
+	sep_priv->ep_fid.fid.ops = &gnix_sep_fi_ops;
+	sep_priv->ep_fid.ops = &gnix_sep_ops;
+	sep_priv->ep_fid.cm = &gnix_cm_ops;
+	sep_priv->domain = domain;
+
+	sep_priv->info = fi_dupinfo(info);
```

Comment:
I think we needed this at one point because of fields in the info struct that we were subsequently using.  If these are no longer needed, then this field can be dropped from ```struct sep_priv```.